typescript/infra/scripts/verify-validator.ts

```diff
@@ -0,0 +1,300 @@
+    .describe('c', 'S3 bucket of the the known (control) validator')
+    .demandOption('c')
+    .string('c').argv;
+}
+
+class S3Wrapper {
+  private readonly client: S3Client;
+  readonly region: string;
+  readonly bucket: string;
+
+  constructor(bucketUrl: string) {
+    const match = bucketUrl.match(
+      /^https:\/\/(.*)\.s3\.(.*)\.amazonaws.com\/?$/,
+    );
+    if (!match) throw new Error('Could not parse bucket url');
+    this.bucket = match[1];
+    this.region = match[2];
+    this.client = new S3Client({ region: this.region });
+  }
+
+  async getS3Obj<T = unknown>(key: string): Promise<T> {
+    const response = await this.client.send(
+      new GetObjectCommand({
+        Bucket: this.bucket,
+        Key: key,
+      }),
+    );
+    if (!response.Body) {
+      throw new Error('No data received');
+    }
+    const bodyStream: NodeJS.ReadableStream =
+      'stream' in response.Body
+        ? response.Body.stream()
+        : (response.Body as NodeJS.ReadableStream);
+
+    const body: string = await streamToString(bodyStream);
+    return JSON.parse(body);
+  }
+
+  async getLastModTime(key: string): Promise<Date> {
+    const request = await this.client.send(
+      new ListObjectsCommand({
+        Bucket: this.bucket,
+        Prefix: key,
+      }),
+    );
+
+    const latestIdx = request.Contents!.findIndex(
+      (v) => v.LastModified && v.Key == key,
+    );
+    const latest = request.Contents![latestIdx]!;
+    return latest.LastModified!;
+  }
+}
+
+async function main() {
+  const {
+    a: _validatorAddress,
+    p: prospectiveBucket,
+    c: controlBucket,
+  } = await getArgs();
+
+  const cClient = new S3Wrapper(controlBucket);
+  const pClient = new S3Wrapper(prospectiveBucket);
+
+  const [cLatestCheckpoint, pLastCheckpoint] = await Promise.all([
+    cClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
+      console.error(
+        "Failed to get control validator's latest checkpoint.",
+        err,
+      );
+      process.exit(1);
+    }),
+    pClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
+      console.error(
+        "Failed to get prospective validator's latest checkpoint.",
+        err,
+      );
+      process.exit(1);
+    }),
+  ]);
+
+  console.assert(
+    Number.isSafeInteger(cLatestCheckpoint),
+    'Expected latest control checkpoint to be an integer',
+  );
+  console.assert(
+    Number.isSafeInteger(pLastCheckpoint),
+    'Expected latest prospective checkpoint to be an integer',
+  );
+
+  console.log(`Latest Index`);
+  console.log(`control: ${cLatestCheckpoint}`);
+  console.log(`prospective: ${pLastCheckpoint}\n`);
+
+  let extraCheckpoints = [];
+  const missingCheckpoints = [];
+  let invalidCheckpoints = [];
+  const modTimeDeltasS = [];
+  const fullyCorrectCheckpoints = [];
+  let missingInARow = 0;
+  for (let i = Math.max(cLatestCheckpoint, pLastCheckpoint); i >= 0; --i) {
+    if (missingInARow == 10) {
```

Comment:
may be nice to pull the 10 out as a constant.

Also - because a checkpoint indices are basically the same thing as message indices, it's possible for many messages to be dispatched in a single block, and the expected behavior is for validators to sign the checkpoint with the index that is the latest index of all the messages dispatched in the block. Although we haven't seen this so far, it's possible for >= 10 messages to be dispatched in a block and this code path would be hit.

So instead of the condition being "if there is ever a situation where a range of 10 checkpoint indices are not signed", I think the more accurate condition would be "if there is a period of 10 checkpoints (whose indices may not necessarily be consecutive) that the control validator has signed, but the prospective has not signed all of them". E.g. if the control validator has signed these indices:
`[100, 85, 83, 78, 75, 73, 70, 67, 65, 63]`, this condition should be hit if the prospective validator hasn't signed all of those indices too, rather than right now where this condition will be hit nearly immediately because `100 - 85` >= 10

Reply:
Done